### ISSUE TITLE
Time window violation

### ISSUE THREAD

**Issue by lukes** [graphhopper/jsprit#281]
Hi, we're finding a situation where it appears as though a time window for service isn't being obeyed.

I've reproduced the full code below. 

The problem is with service job "11790". We set its time window to `1474495200000`, however the arrTime for this job ends up being set to `1474488300000`.

```java
VehicleRoutingProblem.Builder vrpBuilder = VehicleRoutingProblem.Builder.newInstance();


// Vehicles

VehicleTypeImpl.Builder vehicleTypeBuilder1 = VehicleTypeImpl.Builder.newInstance("type_2_0");
vehicleTypeBuilder1.addCapacityDimension(0, 999999999);
vehicleTypeBuilder1.setFixedCost(0);
vehicleTypeBuilder1.setCostPerDistance(0);

VehicleTypeImpl vehicleType1 = vehicleTypeBuilder1.build();
VehicleImpl.Builder vb1 = VehicleImpl.Builder.newInstance("2_0");
vb1.setType(vehicleType1);
vb1.setStartLocation(Location.newInstance(174.780233, -36.865421));
vb1.setReturnToDepot(false);
vb1.setEarliestStart(1474488000000L); // "2016-09-21T20:00:00Z"
vb1.setLatestArrival(1474524000000L); // "2016-09-22T06:00:00Z"
vrpBuilder.addVehicle(vb1.build());


VehicleTypeImpl.Builder vehicleTypeBuilder2 = VehicleTypeImpl.Builder.newInstance("type_2_1");
vehicleTypeBuilder2.addCapacityDimension(0, 999999999);
vehicleTypeBuilder2.setFixedCost(0);
vehicleTypeBuilder2.setCostPerDistance(0);

VehicleTypeImpl vehicleType2 = vehicleTypeBuilder2.build();
VehicleImpl.Builder vb2 = VehicleImpl.Builder.newInstance("2_1");
vb2.setType(vehicleType2);
vb2.setStartLocation(Location.newInstance(174.780233, -36.865421));
vb2.setReturnToDepot(false);
vb2.setEarliestStart(1474574400000L); // "2016-09-22T20:00:00Z"
vb2.setLatestArrival(1474610400000L); // "2016-09-23T06:00:00Z"
vrpBuilder.addVehicle(vb2.build());


// Services
Service service1 = Service.Builder.newInstance("11787")
    .addSizeDimension(0, 1)
    .setServiceTime(300000)
    .setLocation(Location.newInstance(174.76451710000003, -36.8508999))
    .setPriority(1).build();
vrpBuilder.addJob(service1);


Service service2 = Service.Builder.newInstance("11788")
    .addSizeDimension(0, 12) 
    .setServiceTime(5400000)
    .setLocation(Location.newInstance(174.7175866, -36.8110651))
    .addTimeWindow(new TimeWindow(1474502400000L, 1474507800000L))
    .setPriority(1).build();
vrpBuilder.addJob(service2);

// This is the one that has its time window violated
Service service3 = Service.Builder.newInstance("11790")
    .addSizeDimension(0, 3) 
    .setServiceTime(10800000)
    .setLocation(Location.newInstance(174.71151839999993, -36.8181182))
    .addTimeWindow(new TimeWindow(1474495200000L, 1474498800000L))
    .setPriority(1).build();
vrpBuilder.addJob(service3);


// Find and print the solution
final VehicleRoutingProblem vrp = vrpBuilder.build();

VehicleRoutingAlgorithm algorithm = Jsprit.createAlgorithm(vrp);
Collection<VehicleRoutingProblemSolution> solutions = algorithm.searchSolutions();
VehicleRoutingProblemSolution best = Solutions.bestOf(solutions);



SolutionPrinter.print(vrp, best, SolutionPrinter.Print.VERBOSE);
```

Which produces this output:

```
+--------------------------+
| problem                  |
+---------------+----------+
| indicator     | value    |
+---------------+----------+
| noJobs        | 3        |
| noServices    | 3        |
| noShipments   | 0        |
| noBreaks      | 0        |
| fleetsize     | INFINITE |
+--------------------------+
+----------------------------------------------------------+
| solution                                                 |
+---------------+------------------------------------------+
| indicator     | value                                    |
+---------------+------------------------------------------+
| costs         | 0.0                                      |
| noVehicles    | 1                                        |
| unassgndJobs  | 0                                        |
+----------------------------------------------------------+
+--------------------------------------------------------------------------------------------------------------------------------+
| detailed solution                                                                                                              |
+---------+----------------------+-----------------------+-----------------+-----------------+-----------------+-----------------+
| route   | vehicle              | activity              | job             | arrTime         | endTime         | costs           |
+---------+----------------------+-----------------------+-----------------+-----------------+-----------------+-----------------+
| 1       | 2_0                  | start                 | -               | undef           | 1474488000000   | 0               |
| 1       | 2_0                  | service               | 11787           | 1474488000000   | 1474488300000   | 0               |
| 1       | 2_0                  | service               | 11790           | 1474488300000   | 1474506000000   | 0               |
| 1       | 2_0                  | service               | 11788           | 1474506000000   | 1474511400000   | 0               |
| 1       | 2_0                  | end                   | -               | 1474511400000   | undef           | 0               |
+--------------------------------------------------------------------------------------------------------------------------------+
```

Our dependencies look like this:

```
apply plugin: 'java'

sourceCompatibility = 1.8

repositories {
    mavenCentral()
    maven { url "https://oss.sonatype.org/content/groups/public/" }
    maven { url "https://github.com/jsprit/mvn-rep/raw/master/snapshots" }
}

configurations.all {
    // Check for updates every build
    resolutionStrategy.cacheChangingModulesFor 0, 'seconds'
}

dependencies {
    testCompile group: 'junit', name: 'junit', version: '4.11'

    compile("jsprit:jsprit-core:1.6.3-SNAPSHOT") { changing = true }
    compile("jsprit:jsprit-analysis:1.6.3-SNAPSHOT") { changing = true }
    compile("jsprit:jsprit-instances:1.6.3-SNAPSHOT") { changing = true }

    compile "com.graphhopper:graphhopper:0.7.0"
    compile "com.graphhopper:graphhopper-tools:0.7.0"

    compile "com.google.guava:guava:19.0"
    compile "com.google.code.gson:gson:2.5"
    compile "com.fatboyindustrial.gson-javatime-serialisers:gson-javatime-serialisers:1.1.1"

    compile "org.apache.commons:commons-lang3:3.4"

    compile "com.rabbitmq:amqp-client:3.6.0"
}
```

**Comment by oblonski** [comment]
Hi, Thanks for reporting. I do not see any time window violation.

jobId: 11790 --> latest: 1474498800000L > arr_time: 1474488300000
jobId: 11788 --> latest: 1474507800000L > arr_time: 1474506000000
end: --> latest: 1474610400000L > arr_time: 1474511400000

Did I miss smth?

**Comment by lukes** [comment]
Hi, thank you very much for your quick reply. 

This is very likely due to us not understanding the API for jsprit properly, and if this is the case then I apologise for taking up your time!

My understanding is that `service3` (jobID 11790) is being assigned a time window, defined as "the earliest time this can be started is `1474495200000`, and the latest time it can be started is `1474498800000`". 

In which case, the generated arrival time of `1474488300000` is earlier than the start of that time window.

What am I missing?

Thanks.

**Comment by PGWelch** [comment]
Hi Luke,


you can arrive at a time window before its open; you just have to wait there for it to open. This is presumably what you're seeing.

**Comment by lukes** [comment]
Hi, 

Ah okay - I understand now. We're using the response slightly differently so misinterpreted what the response was meaning. We're using jsprit to provide us with optimal start times and end times for jobs within given windows of time. In our case, if the arrival time is earlier than the time window, we'll set the job's start time to when the time window starts, and it should still work fully.

Many thanks!

I've found there's been a few gotchas with interpreting the example code, and wonder if there's a documented API of the class methods somewhere? For example, I discovered that the end of the time window means "earliest start time", rather than "end time", which is what I expected, by chance on Stack Overflow. However - immeasurable thanks for maintaining this incredible library!

Luke
